Improve NXP CSI and  MIPI_CSI2Rx drivers

ngphibang opened the pull request:
This PR improves the NXP's csi and mipi_csi2_rx drivers including
- Some minor fixes & optimization
- Support for changing frame rate
- **Fixed the [chicken-egg issue on NXP i.MX RT10xx](https://github.com/zephyrproject-rtos/zephyr/issues/80304)**

Fixes #80304

This PR **depends on** ~~#74415 and #72254, also **relates to** #76144~~

ngphibang committed:
drivers: video: mipi_csi2rx: Get data lanes number from devicetree

Get number of data lanes from device tree instead of hard-coding it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

0xFarahFl committed:
drivers: video: mipi_csi2rx: Add set_ctrl callback

Add set_ctrl callback to propagate controls to the sensor.

Signed-off-by: Farah Fliss <farah.fliss@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: csi: Add support for changing frame rate

Add support for changing frame rate

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: mipi_csi2rx: Add support for changing frame rate

Add support for changing frame rate.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Remove obsolete comment

Remove the obsolete comment about the init order of the CSI and the
camera sensors (e.g. mt9m114) which is not true anymore.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Add NXP copyright

The CSI is an NXP IP and the driver has been very much changed
by NXP, so add NXP copyright to it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp committed:
drivers: video: mipi_csi2rx: Set clocks according to pixel rate

Instead of fixing csi2rx clock frequencies, set them according to the
pixel rate got from the camera sensor.

Signed-off-by: Trung Hieu Le <trunghieu.le@nxp.com>
Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

trunghieulenxp commented:
Rebase because of the PR https://github.com/zephyrproject-rtos/zephyr/pull/73009 is merged.

trunghieulenxp commented:
Rebase because the PR https://github.com/zephyrproject-rtos/zephyr/pull/76393 is merged

danieldegrasse commented:
> Rebase because the PR https://github.com/zephyrproject-rtos/zephyr/pull/76393 is merged

@trunghieulenxp while you are welcome to rebase as much as you please, it may not be worth bothering to do so until the dependent PRs are merged, especially if CI will fail until that point. If a change is large I'll often review it before CI is passing, but in this case the change is primarily at the driver level, so it likely won't generate a ton of engagement until CI can pass.

ngphibang commented:
Updated according changes in https://github.com/zephyrproject-rtos/zephyr/pull/74415.

@danieldegrasse : Yes, but since there are new changes comming, we update it regularly to not forget things :). Another reason is to cooperate with [this PR](https://github.com/zephyrproject-rtos/zephyr/pull/76124).

ngphibang committed:
dts-bindings: video: ov7725: Use video interfaces binding

Switch to use the new video interfaces binding

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: ov5640: Use video interfaces binding

Switch to use the new video interfaces binding

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: mipicsi2rx: Use video interfaces bindings

Switch to use the new video interfaces bindings

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
dts-bindings: video: csi: Use video interfaces bindings

Switch to use the new video interfaces bindings

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
boards: shields: nxp_btb44_ov5640: Add some endpoint properties

The ov5640 camera driver now supports both MIPI CSI2 (DPHY) and DVP
modes. It is in MIPI CSI2 mode in this overlay. Add bus-type property
for this. In this mode, data-lanes property is required as well.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
As dependencies are now merged, could you help to review this to move this PR along ? @danieldegrasse

ngphibang commented:
CI still failed, let me fix it. Sorry

josuah commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -182,12 +164,136 @@ static int mipi_csi2rx_get_caps(const struct device *dev, enum video_endpoint_id
+		return ret;
+	}
+
+	ret = video_get_ctrl(config->sensor_dev, VIDEO_CID_PIXEL_RATE, (uint64_t *)&cur_pixel_rate);
```
I suppose all other sensors would need a `VIDEO_CID_PIXEL_RATE` or they would not work with `CSI2RX`?
I added it [there](https://github.com/zephyrproject-rtos/zephyr/pull/79482/files#diff-49b6bd99cd52e3fdc82d995c918f790db16e8e9a3fc5f329a8dbb2edf4315ab7R243-R245)

ngphibang replied:
`VIDEO_CID_PIXEL_RATE` is only needed if you want to support changing framerate. And it's only one way to do, you can also use link frequencies.

josuah commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -182,12 +164,136 @@ static int mipi_csi2rx_get_caps(const struct device *dev, enum video_endpoint_id
+		return ret;
+	}
+
+	ret = video_get_ctrl(config->sensor_dev, VIDEO_CID_PIXEL_RATE, (uint64_t *)&cur_pixel_rate);
```
So I suppose the de-facto standard for CID types is now to look at [Linux source](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/media/v4l2-core/v4l2-ctrls-defs.c#n1516)?

Not a bad idea as it permits to always have a reference and it had plenty of time to get stable and cover all use-cases.

Another PR could discuss which API to get a control size (0 for immediate cast or >0 for size of type pointed) or type.

josuah reviewed (dismissed):
Good to have an example of how the frame rate can be computed from inside a driver!

The comments are mostly discussions about what comes next, not issues about this driver.

ngphibang commented on dts/bindings/video/nxp,imx-csi.yaml:
```diff
@@ -19,3 +19,7 @@ properties:
     description: the connected source device,
       e.g., a mipi csi or a camera sensor
+
+child-binding:
+  child-binding:
+    include: video-interfaces.yaml
```
Yes, changing to the new video interfaces bindings requires changes in several platform .dts, test cases which is difficult to cover all at once, so that's why I said, it's better to let the owners do it but I don't know how to inform people about it.

josuah replied:
Would opening an issue help? One to keep track of the propagation of this new APIs through the various video drivers, and flagging the appropriate teams.

ngphibang replied:
yes, will do.

josuah replied:
I hope it matches what you had in mind... I opened this:
- https://github.com/zephyrproject-rtos/zephyr/pull/76475

ngphibang replied:
Thank you very much for creating the issue :-). I think you've mistaken the [issue link](https://github.com/zephyrproject-rtos/zephyr/issues/80514) though. I made some modifications on the issue description.

josuah replied:
Oops! Yes... Thank you. I created it with the "bug" template but I agree.

ngphibang commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -182,12 +164,136 @@ static int mipi_csi2rx_get_caps(const struct device *dev, enum video_endpoint_id
+		return ret;
+	}
+
+	ret = video_get_ctrl(config->sensor_dev, VIDEO_CID_PIXEL_RATE, (uint64_t *)&cur_pixel_rate);
```
yes, changed it.

ngphibang reviewed (commented):
Force push to fix CI issue which is due to `tests/build_all/video` failed due to the changes to the new video interfaces binding not taken into account

josuah commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -182,12 +163,136 @@ static int mipi_csi2rx_get_caps(const struct device *dev, enum video_endpoint_id
+		return ret;
+	}
+
+	ret = video_get_ctrl(config->sensor_dev, VIDEO_CID_PIXEL_RATE, &cur_pixel_rate);
+	if (ret) {
+		LOG_ERR("Cannot get sensor_dev pixel rate");
+		return ret;
+	}
```
Maybe there is an inconsistency bewteen the `video_get_ctrl()` API and the `_enum_frmival()` API as used here?

It does math on the [*current* pixel rate](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/ov5640.c#L376-L380) and the *queried* format from `fie` (which is not applied yet).

Would that allow a pixel rate for a wrong format to be used in the frame interval math?
Should the pixel rate become part of the `fie` query struct so that it can be filled by `sensor_dev`?
Should the `fie` not contain any format and the current format always be used to simplify this situation?

It is very possible that I missed something too...

ngphibang replied:
- the current pixel rate is coming from the sensor and is only updated whenever there is a change in frame rate, so when enum_frmival is called, it always uses the latest / updated value.

> - Should the fie not contain any format and the current format always be used to simplify this situation?

=> No, fie is used to query supported framerates for _any_ given format, not only the current one. Thats why it must contain the format in the query.

josuah commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -7,19 +7,18 @@
-#define DEFAULT_CAMERA_FRAME_RATE  30
+#define MAX_SUPPORTED_PIXEL_RATE MHZ(96)
+
+#define ABS(a, b) (a > b ? a - b : b - a)
```
The `ABS()` macro is sometimes defined as `a > 0 ? a : -a` (like C's [`abs()`](https://www.man7.org/linux/man-pages/man3/abs.3.html) from `<stdlib.h>`) 
I proposed an `ABS_DIFF(a, b)` in `util.h` [there](https://github.com/zephyrproject-rtos/zephyr/pull/79482/files#diff-c5b30d0a3cf86108357ba5a1c2a12f48f67d2f375e2adbe7549296f3c720f52d).

I do not think there is an issue with having it defined in `video_mcux_mipi_csi2rx.c` as it is not an `#include` file so no side effect.

ngphibang committed:
dts-bindings: video: mt9m114: Use video interfaces binding

Switch to use the new video interfaces binding

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

dleach02 commented:
@josuah if you are good with these changes can you give it an approval

josuah reviewed (dismissed):
@dleach02 yes my bad, I now understood that the `video_frmival_enum()` API is listing FPS for the *queried* format and the *current* CIDs: not a bug.

danieldegrasse commented on drivers/video/video_mcux_csi.c:
```diff
@@ -32,25 +32,6 @@ struct video_mcux_csi_data {
 	struct k_poll_signal *signal;
 };
 
-static inline unsigned int video_pix_fmt_bpp(uint32_t pixelformat)
```
It seems like the `stm32_dcmi` driver also has a definition like this- should we clean that up here too?

danieldegrasse commented on drivers/video/video_mcux_csi.c:
```diff
@@ -198,6 +198,11 @@ static int video_mcux_csi_stream_start(const struct device *dev)
 	status_t ret;
+	struct video_buf *vbuf;
+
+	while ((vbuf = k_fifo_get(&data->fifo_out, K_NO_WAIT))) {
```
I'm unclear what the goal here is. If a consumer is using the video API correctly, they should queue at least `min_vbuf_count` buffers prior to calling `video_stream_start`, right? So the video pipeline should be able to start. Why wouldn't we simply discard any buffers in the `fifo_out` queue when `video_stream_stop` was called to avoid the issue with dequeuing?

danieldegrasse replied:
Ok, I see why we have this requirement, but it seems like a *software* requirement, not a hardware one. The HAL layer chooses to enqueue all remaining buffers, but we could simply bypass calling `CSI_TransferStop`, and call `CSI_Stop` instead, right? My concern is that a user of the video API won't expect buffers that were previously queued to be reused after the video stream is restarted, will they? We need to follow the video API here.

trunghieulenxp replied:
- With the LL CSI driver, bypass CSI_TransferStop will not resolve the problem as low-level driver can still hold an in-progress buffer.

>  My concern is that a user of the video API won't expect buffers that were previously queued to be reused after the video stream is restarted, will they? We need to follow the video API here.

- As my observation, in video API, user is expecting that the buffer is reused. This is because when user call video_stream_stop function, all buffers from the input queue will **be flushed** (https://github.com/zephyrproject-rtos/zephyr/blob/58d503268573d465b6f8f456190930e06eef594f/include/zephyr/drivers/video.h#L574)  to the **output_fifo** (https://github.com/zephyrproject-rtos/zephyr/blob/aefda52e42b599592dcfbcbd359aa8dd6e908994/drivers/video/video_mcux_csi.c#L271) , and user must dequeue these buffers from the output fifo. So, logically, if the user is not dequeuing all the buffers in the output fifo, remaining buffers must be moved back to the input fifo to prevent unexpected behavior.

danieldegrasse commented on drivers/video/video_mcux_csi.c:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 2019, Linaro Limited
+ * Copyright 2024 NXP
```
This can be in the first commit where we touch the CSI file in this PR

ngphibang replied:
In fact, we touched the CSI driver several times in the past and I explictly added the NXP Copyright here when seeing that the driver is changed enough by NXP (although I don't know how much change is "enough" to add a copyright).

danieldegrasse commented on boards/shields/nxp_btb44_ov5640/nxp_btb44_ov5640.overlay:
```diff
@@ -21,7 +21,7 @@
 
 		port {
 			ov5640_ep_out: endpoint {
-				remote-endpoint = <&mipi_csi2rx_ep_in>;
```
Just to make sure I understand here, is the reason we want to use `remote-endpoint-label` over `remote-endpoint` that it avoids the "check-and-egg" dependency problem we have with display and camera drivers? I primarily ask because this deviates from how Linux describes CSI devices in devicetree, right?

ngphibang replied:
If the question is for this PR, we need the new video-interfaces bindings for the mipi csi2rx to be able to use the data-lanes property and for the ov5640 to be able to use the bus-type (and data-lanes) property as the driver now supports both CSI2 and DVP modes.

If the question is for the video-interfaces binding (which is merged), replacing remote-endpoint (phandle) by remote-endpoint-label (string) is mandatory as we cannot introduce any binding for endpoint without doing this because it simply does not compile (due to circular dependency).

> I primarily ask because this deviates from how Linux describes CSI devices in devicetree, right?

Apart from the above reason, remote-endpoint-label also helps to bypass the technical obstacle in Zephyr (circular dependency) **to keep Zephyr does NOT deviates from Linux**. In Linux, they don't directly reference the peer devices in DT with direct phandle like we did in Zephyr (with "`source-dev`", "`sensor-dev`", etc). All is done through `remote-endpoint` (some advantages : it's bidirectional, can connect elements at runtime, etc.). And now we can do the same in Zephyr with `remote-endpoint-label`. Someday in the future, if Zephyr supports circular dependency (`remote-endpoint`), we can switch from `remote-endpoint-label` to `remote-endpoint` easily as all the mechanisms will remain the same (it's much easier than the direct phandle reference with `source-dev`, `sensor-dev` we did today).

danieldegrasse commented on tests/drivers/build_all/video/mimxrt1170_evk_mimxrt1176_cm7.overlay:
```diff
@@ -8,6 +8,8 @@
  * (and be extended to test) real hardware.
  */
 
+#include <zephyr/dt-bindings/video/video-interfaces.h>
```
Test changes need to be with the change to the binding to preserve bisectability

danieldegrasse commented on boards/shields/nxp_btb44_ov5640/nxp_btb44_ov5640.overlay:
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <zephyr/dt-bindings/video/video-interfaces.h>
```
Same thing here, these test changes need to be in the same commit as the binding changes

danieldegrasse commented on drivers/video/video_mcux_mipi_csi2rx.c:
```diff
@@ -25,98 +28,82 @@ struct mipi_csi2rx_data {
-	drv_data->csi2rxConfig = csi2rxConfig;
+	sensor_byte_clk = sensor_pixel_rate * bpp / drv_data->csi2rxConfig.laneNum / 8;
+	if (sensor_byte_clk > CLOCK_GetRootClockFreq(kCLOCK_Root_Csi2)) {
+		mipi_csi2rx_clock_set_freq(kCLOCK_Root_Csi2, sensor_byte_clk);
```
There should be one more layer of abstraction here- you should be calling `clock_control_set_rate` on the clock device here. See what is done for the FlexSPI driver:
https://github.com/zephyrproject-rtos/zephyr/blob/a70eaa5c4a65c24c9ad5115df62b0905d5557723/drivers/memc/memc_mcux_flexspi.c#L106

trunghieulenxp replied:
Thanks, fixed it with your suggestions.

ngphibang committed:
drivers: video: mipi_csi2rx: Drop sensor device phandle reference

The peer remote device "sensor_dev" can be retrieved from the
remote-endpoint-label. Direct reference via phandle is not needed.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Drop source device phandle reference

The peer remote device "source_dev" can be retrieved from the
remote-endpoint-label. Direct reference via phandle is not needed.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: csi: Increase init priority

The CSI needs to be initialized BEFORE the camera sensor to provide
clock to the camera sensor. It is now possible to do so as direct
reference to the sensor via phandle is now removed. There will be
no check failure on the init order anymore when compiling.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -827,6 +827,30 @@ void video_buffer_release(struct video_buffer *buf);
+ *
+ * @param pixfmt FourCC pixel format value (\ref video_pixel_formats).
+ */
+static inline unsigned int video_pix_fmt_bpp(uint32_t pixfmt)
+{
+	switch (pixfmt) {
+	case VIDEO_PIX_FMT_BGGR8:
+	case VIDEO_PIX_FMT_GBRG8:
+	case VIDEO_PIX_FMT_GRBG8:
+	case VIDEO_PIX_FMT_RGGB8:
+		return 1;
+	case VIDEO_PIX_FMT_RGB565:
+	case VIDEO_PIX_FMT_YUYV:
+		return 2;
+	case VIDEO_PIX_FMT_XRGB32:
+	case VIDEO_PIX_FMT_XYUV32:
+		return 4;
+	default:
+		return 0;
+	}
+}
```
I proposed a `video_bits_per_pixel()` here: https://github.com/zephyrproject-rtos/zephyr/pull/79476
But it is even better if a PR on improving the driver safely moves the utility function into a the shared source.

A `video_byte_per_pixel()` cannot support 10-bit or 12-bit formats? These are well-standardized by i.e. MIPI which defines the bit packing.

If using to mean "bits per pixel", this would work as the multiplication is done before the division:
```c
fmt.pitch = fmt.width * video_bits_per_pixel(fmt.pixelformat) / 8;
```

![scrot_20241023_122633_778x229](https://github.com/user-attachments/assets/b90937b5-68cf-45bb-b85f-716358e91744)

ngphibang replied:
Yes, I am aware of your PR and agree it is better. Here, we will use this helper function in several places in this PR to change frame rate so a quick refactor / cleanup is needed unless reviewers will say "code duplication".

Feel free to change this if this PR is merged first or I can rebase our PR if yours is merged first.

josuah reviewed (commented):
Just bringing attention on some detail on the utility in `video.h`, making sure this was not glossed over. It does not matter if it is not adjusted *in this PR*, as it is easy to make another PR to propose the modification.

These improvement will help building the [emulated video DMA RX driver](https://github.com/zephyrproject-rtos/zephyr/pull/79482/files#diff-af6e1048ac91cb76bf66f5ef9de7a1601cb7770ae92496b4dc7b1b3dae45d835), which I will refactor to match the style of this driver.

ngphibang committed:
include: video: Add an utility function to get bytes per pixel

As getting bytes per pixel of a pixel format is a very common operation,
add an utility function for it instead of repeating the same codes in
different drivers.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
> I proposed a video_bits_per_pixel() here: https://github.com/zephyrproject-rtos/zephyr/pull/79476

@josuah : yes, I am aware of your PR and agree it's better. Here, I don't intend to do a "perfect" utility function for video.h. It 's just because we will use this helper functions in several places (for frame rate changing) so I have to do a quick refactor / cleanup unless reviewers will say "code dupplication" :-).

Feel free to change this if this PR is merged before your PR or I will rebase our PR if yours is merged first :-).

ngphibang reviewed (commented):
Force push to address comments from reviewers. Some are still in progress and will be pushed later by @trunghieulenxp .
Changed PR description

danieldegrasse reviewed (dismissed):
Change request is specifically about the choice to reuse vbufs previously supplied by the user when we start streaming with the CSI- based on how I understand the video API works, we should clear out the buffers in the CSI when the user stops a stream, not reuse them. If the HAL layer does not currently enable this, we need to either:
- bypass the HAL layer and interact with the CSI directly
- Modify the `handle` data structure used by the HAL directly to discard the queued buffers

ngphibang commented:
@danieldegrasse Regarding the commit "**drivers: video: csi: Fix start / stop issue**", your points are almost right. To summerize:

- After stopping, users are expected to re-enqueue the buffers before restart streaming
- However, when stopping, the SDK CSI driver puts all the active buffers into the input (empty) queue with expect that user can start again right away as can be seen [here](https://github.com/zephyrproject-rtos/hal_nxp/blob/99570ee6fa32a9ab413b8ee9c0226cda26f6a2ae/mcux/mcux-sdk/drivers/csi/fsl_csi.c#L705)
- While the Zephyr CSI driver puts all the buffers in the input queue into the output queue as can be seen [here](https://github.com/zephyrproject-rtos/zephyr/blob/aefda52e42b599592dcfbcbd359aa8dd6e908994/drivers/video/video_mcux_csi.c#L271)
- So, to fix the issue with these conflict behaviors, I think we need to call CSI_STOP() and disable interrupts directly without passing by CSI_TransferStop() as you said. The input and output queues in the Zephyr CSI driver need to be clean too.

Given the strict timeline approaching and to facilitate the review, we **removed this commit from the PR** and wait for next time to submit it.

danieldegrasse commented:
> @danieldegrasse Regarding the commit "**drivers: video: csi: Fix start / stop issue**", your points are almost right. To summerize:
> 
>     * After stopping, users are expected to re-enqueue the buffers before restart streaming
> 
>     * However, when stopping, the SDK CSI driver puts all the active buffers into the input (empty) queue with expect that user can start again right away as can be seen [here](https://github.com/zephyrproject-rtos/hal_nxp/blob/99570ee6fa32a9ab413b8ee9c0226cda26f6a2ae/mcux/mcux-sdk/drivers/csi/fsl_csi.c#L705)
> 
>     * While the Zephyr CSI driver puts all the buffers in the input queue into the output queue as can be seen [here](https://github.com/zephyrproject-rtos/zephyr/blob/aefda52e42b599592dcfbcbd359aa8dd6e908994/drivers/video/video_mcux_csi.c#L271)
> 
>     * So, to fix the issue with these conflict behaviors, I think we need to call CSI_STOP() and disable interrupts directly without passing by CSI_TransferStop() as you said. The input and output queues in the Zephyr CSI driver need to be clean too.
> 
> 
> Given the strict timeline approaching and to facilitate the review, we **removed this commit from the PR** and wait for next time to submit it.

Reading the video driver documentation, I think the behavior for `video_mcux_csi_flush` is correct- if the user opts to cancel buffer processing, then queued buffers are supposed to be moved through the queue without being processed (like the implementation does)

There isn't clear documentation on how `video_stream_stop` is supposed to function regarding input buffers. We could either:
- discard all buffers in the input queue, and require user to re-queue them
- keep buffers queued in the input queue, and when the stream is restarted use those buffers

@loicpoulain do you know which behavior is correct?

I do not think we should flush or modify the output buffer queue when the stream is stopped.

ngphibang commented:
> keep buffers queued in the input queue, and when the stream is restarted use those buffers

=> The CSI needs at least 2 buffers to be able to start. Remaining buffers in the input queue may be insufficient. That's why users need to re-queue buffer (option 1).

> I do not think we should flush or modify the output buffer queue when the stream is stopped.

=> If users re-queue (2) buffers (of course they will not alloc new buffers but re-use the ones in the buffer pool). We need to clean the output queue unless buffers will be present on both input and output queue

danieldegrasse commented:
> => The CSI needs at least 2 buffers to be able to start. Remaining buffers in the input queue may be insufficient. That's why users need to re-queue buffer (option 1).

Sorry, I think I might have been unclear. My understanding is that user might interact with the video API like so:

- Queue buffer A
- Queue buffer B
- Queue buffer C
- start CSI stream
- dequeue processed buffer A
- stop CSI stream

At this point, if the CSI has processed buffer B we should expect that buffer to be in the output queue, and if the user tries to dequeue a video buffer they will get buffer B.

Video buffer C is where I am not sure. I *think* that the buffer should remain in the input queue, so that if the user then does something like so:
- Queue buffer A
- Queue buffer B
- dequeue buffer

They would receive buffer C. However, I'm not sure if this is correct, or if we should be clearing the input queue so that the user now would receive buffer A when they run the dequeue operation.

Does this help clarify my question here?

ngphibang commented:
> Video buffer C is where I am not sure. I think that the buffer should remain in the input queue, so that if the user then does something like so:
> 
> Queue buffer A
> Queue buffer B
> dequeue buffer
> They would receive buffer C. However, I'm not sure if this is correct, or if we should be clearing the input queue so that the user now would receive buffer A when they run the dequeue operation.

Yes, buffer C is in the input queue. But I don't really get your idea behind when raising these points ? So, let me summerize what's going on in the application:

- allocate an array of 2 buffers (most of the cases the min_buffer_count = 2)
- enqueue 2 buffers (step X)
- start stream
- In an infinite loop, doing:
=> dequeue a buffer
=> do something with the buffer
=> re-enqueue the buffer 

At this point, user can "stop stream" and this can happen at any steps in the loop. So, the questions are:

- What should user do to restart ?
- What should the (CSI) driver is supposed to do when stopping so that user can restart again ?

My points are

- A restart should be treated as a "new" start. That means users will enqueue the 2 buffers (step X) and call "start stream". Users do not need to know which buffers are in the input or output queue and do not need to dequeue all the buffers remaining in the output queue.
- Currently, at stooping step, when flushing, the Zephyr CSI driver move all the buffers in the input queue into the output queue. A (little) advantage of this is users can get the remaining processed buffer (usually just 1 buffer, the last frame). But the drawback is that **users need to dequeue all the buffers in the output queue to empty this queue before enqueuing them again into the input queue** (while he/she may not do this). That's why I said that we need to modify this to just clean up the input and output queue, so that users can re-enqueue the buffers without any problems.

A typical example of capturing application can be found [here](https://www.kernel.org/doc/html/v4.9/media/uapi/v4l/capture.c.html) in Linux. They don't really do a restart but we can suppose that to restart, just call `start_capturing()` again. And you can see the `start_capturing()` function including the enqueuing buffer steps as I mentionned above.

```
        open_device();
        init_device();
        start_capturing();
        mainloop();
        stop_capturing();
```
